include/verilated_vpi.cpp

```diff
@@ -2384,122 +2422,210 @@ static void vl_strprintf(std::string& buffer, char const* fmt, ...) {
     va_end(args);
 }
 
-void vl_get_value(const VerilatedVar* varp, void* varDatap, p_vpi_value valuep,
-                  const char* fullname) {
+template<typename T>
+struct VarAccessInfo final {
+    size_t bitOffset;
+    size_t wordOffset;
+    T mask_lo, mask_hi;
+    T* ptr;
+};
+
+template<typename T>
+VarAccessInfo<T> vl_var_access_info(const VerilatedVpioVarBase* vop, size_t bitCount, size_t addOffset) {
+    size_t wordBits = sizeof(T) * 8;
+    uint32_t varBits;
+
+    // find how many bits does the variable have
+    if (vop->indexedDim() + 1 < vop->varp()->udims())
+        varBits = vop->varp()->entBits();
+    else
+        varBits = vop->size();
+
+    if (vop->varp()->vltype() == VLVT_REAL)
+        varBits *= sizeof(double) * 8;
+
+    // make sure we're not trying to write outside var bounds
+    assert(varBits > addOffset);
+    bitCount = std::min(bitCount, varBits - addOffset);
+
+    VarAccessInfo<T> info;
+    info.ptr = reinterpret_cast<T*>(vop->varDatap());
+    if (vop->varp()->vltype() == VLVT_WDATA) {
+        assert(sizeof(T) == sizeof(EData));
+        assert(bitCount <= wordBits);
+        info.wordOffset = (vop->bitOffset() + addOffset) / wordBits;
+        info.bitOffset = (vop->bitOffset() + addOffset) % wordBits;
+        if (bitCount + info.bitOffset <= wordBits) {
+            // within single word
+            if (bitCount == wordBits)
+                info.mask_lo = ~(T)0;
+            else
+                info.mask_lo = ((T)1 << bitCount) - 1;
+            info.mask_lo = info.mask_lo << info.bitOffset;
+            info.mask_hi = 0;
+        } else {
+            // straddles word boundary
+            info.mask_lo = ((T)1 << (wordBits - info.bitOffset)) - 1;
```

Comment:
```suggestion
            info.mask_lo = (static_cast<T>(1) << (wordBits - info.bitOffset)) - 1;
```

Replace all other C casts.